homeassistant/components/tuya/__init__.py

```diff
@@ -0,0 +1,296 @@
+                tuya_mq.add_message_listener(device_manager.on_message)
+
+        def remove_device(self, device_id: str):
+            _LOGGER.debug("tuya remove device:%s", device_id)
+            remove_hass_device(hass, device_id)
+
+    __listener = DeviceListener()
+    hass.data[DOMAIN][TUYA_MQTT_LISTENER] = __listener
+    device_manager.add_device_listener(__listener)
+    hass.data[DOMAIN][TUYA_DEVICE_MANAGER] = device_manager
+
+    # Clean up device entities
+    await cleanup_device_registry(hass)
+
+    _LOGGER.debug("init support type->%s", TUYA_SUPPORT_HA_TYPE)
+
+    for platform in TUYA_SUPPORT_HA_TYPE:
+        _LOGGER.debug("tuya async platform-->%s", platform)
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+        hass.data[DOMAIN][TUYA_SETUP_PLATFORM].add(platform)
+
+    return True
+
+
+async def cleanup_device_registry(hass: HomeAssistant):
+    """Remove deleted device registry entry if there are no remaining entities."""
+
+    __device_registry = device_registry.async_get(hass)
+    device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+
+    for dev_id, device_entity in list(__device_registry.devices.items()):
+        for item in device_entity.identifiers:
+            if DOMAIN == item[0] and item[1] not in device_manager.device_map:
+                __device_registry.async_remove_device(dev_id)
+                break
+
+
+def remove_hass_device(hass: HomeAssistant, device_id: str):
+    """Remove device from hass cache."""
+    __device_registry = device_registry.async_get(hass)
+    for entity in list(__device_registry.devices.values()):
+        print(f"inde->>{entity.identifiers}")
+        if device_id in list(entity.identifiers)[0]:
+            __device_registry.async_remove_device(entity.id)
+
+
+async def async_setup(hass, config):
+    """Set up the Tuya integration."""
+    tuya_logger.setLevel(_LOGGER.level)
+    conf = config.get(DOMAIN)
+
+    _LOGGER.debug("Tuya async setup conf %s", conf)
+    if conf is not None:
+
+        async def flow_init() -> Any:
+            try:
+                result = await hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+                )
+            except UnknownFlow as flow:
+                _LOGGER.error(flow.args)
+            except UnknownStep as step:
+                _LOGGER.error(step.args)
+            except ValueError as err:
+                _LOGGER.error(err.args)
+            _LOGGER.debug("Tuya async setup flow_init")
+            return result
+
+        hass.async_create_task(flow_init())
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unloading the Tuya platforms."""
+    _LOGGER.debug("integration unload")
+    unload = await hass.config_entries.async_unload_platforms(
+        entry, hass.data[DOMAIN]["setup_platform"]
```

Comment:
Unload all the platforms for which you initiated setup

```suggestion
        entry, TUYA_SUPPORT_HA_TYPE
```